### PR TITLE
Add new formatting options by replacing `d3.format` method with more recent `d3-format` module

### PR DESCRIPTION
**Pull request by archmoj**
An attempt to replace old d3 format functions with latest `d3-number` module as part of d3 bump cc: #424; while keep trimming the default option which was quite tricky.

TODO:
 - [x] test more cases to ensure old options still work
 - [x] test new options
 - [x] add fall back for the edge cases
 - [x] drop `d3.format` from `plotly/d3` fork and use it here 

Also to note:  
 - [x] dropped `d3.time` from `plotly/d3` fork and switched to `d3-time` module in https://github.com/plotly/plotly.js/pull/5125/commits/dc0d9cb2433cd4ad1dbb7c1b1612a825b5e72c33 cc: #424 

@plotly/plotly_js

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Cool! Does this sidestep the challenges encountered in https://github.com/plotly/plotly.js/pull/5023 ?

**Comment by archmoj** [comment]
> Cool! Does this sidestep the challenges encountered in #5023 ?

Positive. #5023 was not successful.

**Comment by archmoj** [comment]
Resolved conflicts and added to [v2.3.0 milestone](https://github.com/plotly/plotly.js/milestone/63).

File: package.json [hunk 1]
```diff
@@ -66,7 +66,7 @@
     ]
   },
   "dependencies": {
-    "@plotly/d3": "3.7.0",
+    "@plotly/d3": "git://github.com/plotly/d3.git#3fc195e30e85b008238783379d2aa067c737e636",
```
Comment (archmoj) [inline package.json:69]: https://github.com/plotly/d3/pull/4

File: src/lib/index.js [hunk 2]
```diff
@@ -2,15 +2,43 @@
...
+lib.adjustFormat = function adjustFormat(formatStr) {
+    if(
+        !formatStr ||
+        /^[0123456789].[0123456789]f/.test(formatStr) ||
+        /.[0123456789]%/.test(formatStr)
+    ) return formatStr;
+
+    if(formatStr === '0.f') return '~f';
+    if(/^[0123456789]%/.test(formatStr)) return '~%';
+    if(/^[0123456789]s/.test(formatStr)) return '~s';
```
Comment (alexcjohnson) [inline src/lib/index.js:24]:
```suggestion
        /^\d[.]\df/.test(formatStr) ||
        /[.]\d%/.test(formatStr)
    ) return formatStr;

    if(formatStr === '0.f') return '~f';
    if(/^\d%/.test(formatStr)) return '~%';
    if(/^\ds/.test(formatStr)) return '~s';
```

ie `[0123456789]` === `\d` and you don't mean `.` (any character), you mean `[.]` (just a period character)

File: src/lib/index.js [hunk 3]
```diff
@@ -1319,6 +1347,10 @@ lib.bigFont = function(size) {
     return Math.round(1.2 * size);
 };
 
+lib.beginCap = function(str) {
+    return str.charAt(0).toUpperCase() + str.slice(1);
```
Comment (alexcjohnson) [inline src/lib/index.js:1351]: https://github.com/plotly/plotly.js/blob/bdb69fc18f3b7787c93fff7ee1db2a634a62ee78/src/lib/index.js#L723-L725

File: src/plots/plots.js [hunk 4]
```diff
@@ -702,7 +703,17 @@ function getFormatter(formatObj, separators) {
...
 
     return {
-        numberFormat: d3.locale(formatObj).numberFormat,
+        numberFormat: function(formatStr) {
+            try {
+                formatStr = formatLocale(formatObj).format(
+                    Lib.adjustFormat(formatStr)
+                );
+            } catch(e) {
+                return Lib.noFormat;
```
Comment (alexcjohnson) [inline src/plots/plots.js:712]: Is this catching invalid format strings? Do we want to `lib.warn` when this happens, to help people fix it?

File: src/lib/index.js (hunk 3 above)
Comment (archmoj) [reply src/lib/index.js]: Done in https://github.com/plotly/plotly.js/pull/5125/commits/1efc41eb012ef653486414a616aa05de3b84f879

**Review by alexcjohnson — APPROVED** [review]
💃 after those couple of comments and merging/publishing https://github.com/plotly/d3/pull/4

Nice job with `lib.adjustFormat`!

File: src/plots/plots.js (hunk 4 above)
Comment (archmoj) [reply src/plots/plots.js:713]: Done in https://github.com/plotly/plotly.js/pull/5125/commits/878034b5ae2e78d5b855d2267325d330e8669715
